ext/timers/01_timers.js

```diff
@@ -61,534 +48,307 @@
+      nestingLevel: timerNestingLevel + 1,
+    };
 
-  function clearGlobalTimeout() {
-    globalTimeoutDue = null;
-    opStopGlobalTimer();
+    // 12. Let completionStep be an algorithm step which queues a global task on
+    // the timer task source given global to run task.
+    // 13. Run steps after a timeout given global, "setTimeout/setInterval",
+    // timeout, completionStep, and id.
+    runAfterTimeout(
+      () => ArrayPrototypePush(timerTasks, task),
+      timeout,
+      cancelRid,
+    );
+
+    return id;
   }
 
-  let pendingEvents = 0;
-  const pendingFireTimers = [];
+  // ---------------------------------------------------------------------------
 
-  /** Process and run a single ready timer macrotask.
-   * This function should be registered through Deno.core.setMacrotaskCallback.
-   * Returns true when all ready macrotasks have been processed, false if more
-   * ready ones are available. The Isolate future would rely on the return value
-   * to repeatedly invoke this function until depletion. Multiple invocations
-   * of this function one at a time ensures newly ready microtasks are processed
-   * before next macrotask timer callback is invoked. */
-  function handleTimerMacrotask() {
-    if (pendingFireTimers.length > 0) {
-      fire(ArrayPrototypeShift(pendingFireTimers));
-      return pendingFireTimers.length === 0;
-    }
-    return true;
-  }
+  /**
+   * @typedef ScheduledTimer
+   * @property {number} millis
+   * @property {() => void} cb
+   * @property {boolean} resolved
+   * @property {ScheduledTimer | null} prev
+   * @property {ScheduledTimer | null} next
+   */
 
-  async function setGlobalTimeout(due, now) {
-    // Since JS and Rust don't use the same clock, pass the time to rust as a
-    // relative time value. On the Rust side we'll turn that into an absolute
-    // value again.
-    const timeout = due - now;
-    assert(timeout >= 0);
-    // Send message to the backend.
-    globalTimeoutDue = due;
-    pendingEvents++;
-    // FIXME(bartlomieju): this is problematic, because `clearGlobalTimeout`
-    // is synchronous. That means that timer is cancelled, but this promise is still pending
-    // until next turn of event loop. This leads to "leaking of async ops" in tests;
-    // because `clearTimeout/clearInterval` might be the last statement in test function
-    // `opSanitizer` will immediately complain that there is pending op going on, unless
-    // some timeout/defer is put in place to allow promise resolution.
-    // Ideally `clearGlobalTimeout` doesn't return until this op is resolved, but
-    // I'm not if that's possible.
-    opStartGlobalTimer(timeout);
-    await opWaitGlobalTimer();
-    pendingEvents--;
-    prepareReadyTimers();
-  }
+  /**
+   * A doubly linked list of timers.
+   * @type { { head: ScheduledTimer | null, tail: ScheduledTimer | null } } */
+  const scheduledTimers = { head: null, tail: null };
```

Comment:
You can get rid of all the null checks if you turn this into a list head:
```js
// ideally give the list head and timer objects the same object shape
// to keep property reads and writes monomorphic
const scheduledTimers = { head: null, tail: null };
scheduledTimers.head = scheduledTimers;
scheduledTimers.tail = scheduledTimers;

function isEmpty() {
  return scheduledTimers.tail === scheduledTimers;
}

function append(timer) {
  timer.prev = scheduledTimers.tail;
  timer.next = scheduledTimers;
  scheduledTimers.tail.next = timer;
  scheduledTimers.tail = timer;
}

function remove(timer) {
  timer.prev.next = timer.next;
  timer.next.prev = timer.prev;
  // timer.prev = timer.next = null;
}
```

Reply:
@andreubotella do you want to address this bit before landing?

Reply:
Would be nice to have now and I'd definitely want to see it addressed eventually but I'm okay with landing this as-is.

Reply:
That seems harder to reason through, especially if we go all the way and make `scheduledTimers` monomorphic, since `scheduledTimers.prev` would be the tail and `scheduledTimers.next` would be the head. And if we keep the property names `head` and `tail`, we'd still need checks to append and remove.

Reply:
If it's good enough for nginx, libuv, linux and the BSDs, then it's certainly good enough for deno. :-)

(Also node - but I'm reasonably certain I was the one who introduced that pattern there.)